test-lib.sh

```diff
@@ -40,37 +62,67 @@ UNSTABLE_TESTS="${UNSTABLE_TESTS:-""}"
 function ct_cleanup() {
   ct_show_resources
   ct_show_results
+  ct_clean_app_images
 
   if [[ -z ${CID_FILE_DIR:-} ]]; then
     echo "The \$CID_FILE_DIR is not set. Container cleaning is to be skipped."
     exit "${TESTSUITE_RESULT:-0}"
   fi;
 
-  echo "Examining CID files in $CID_FILE_DIR"
+  # return 0 if container exists
+  function container_exists() {
+    local exists
+    exists="$(docker ps -q -f "id=$1")"
+    [ -n "$exists" ] || return 1
+  }
+
+  echo "Examining CID files in \$CID_FILE_DIR=$CID_FILE_DIR"
   for cid_file in "$CID_FILE_DIR"/* ; do
     [ -f "$cid_file" ] || continue
     local container
     container=$(cat "$cid_file")
+    rm "$cid_file"
 
     echo "Stopping and removing container $container..."
+    container_exists "$container" || continue
     docker stop "$container"
 
     # Container has not been removed by `docker stop` and still exists
-    if [ "$( docker ps -a -f "id=$container" | wc -l )" -eq 2 ]; then
+    if container_exists "$container"; then
       exit_status=$(docker inspect -f '{{.State.ExitCode}}' "$container")
       if [ "$exit_status" != "$EXPECTED_EXIT_CODE" ]; then
         echo "Dumping logs for $container"
         docker logs "$container"
       fi
       docker rm -v "$container"
     fi
-    rm "$cid_file"
   done
 
   rmdir "$CID_FILE_DIR"
   exit "${TESTSUITE_RESULT:-0}"
 }
 
+# ct_clean_app_images
+# --------------------
+# Cleans up application images referenced by APP_ID_FILE_DIR
+# Uses: $APP_ID_FILE_DIR - path to directory containing cid_files
+function ct_clean_app_images() {
+  local image
+  if [[ ! -d ${APP_ID_FILE_DIR:-} ]]; then
+    echo "The \$APP_ID_FILE_DIR=$APP_ID_FILE_DIR is not created. App cleaning is to be skipped."
+    return 0
+  fi;
+  echo "Examining image ID files in \$APP_ID_FILE_DIR=$APP_ID_FILE_DIR"
+  for file in "${APP_ID_FILE_DIR:?}"/*; do
```

Comment:
I don't follow why is there ':?'. Based on the https://devhints.io/bash it means show error message but message is missing and exit if $APP_ID_FILE_DIR is unset.
This is not needed. Variable is defined in ct_init.

Reply:
This is the output if message is not set:
```
$ echo ${APP_ID_FILE_DIR:?}
bash: APP_ID_FILE_DIR: parameter null or not set
```
I thought it could be good to have this check for cases when the `ct_init` hasn't been called for some reason.